markdown/docs/addon-author-guide.md

```diff
@@ -0,0 +1,321 @@
+code to be run, consider using the
+[fastboot-filter-initializers](https://github.com/ronco/fastboot-filter-initializers)
+Broccoli plugin. This allows you to define target-specific
+initializers just by putting them in the right directory.
+
+For example, to write an instance initializer that only runs in the
+FastBoot environment, you'd simply put it in
+`app/instance-initializers/fastboot`.
+
+## Requiring Node Modules
+
+Some addons may need functionality when running in FastBoot that is
+normally provided by the browser. Addons can load Node modules (either
+built-in or from npm) when running in FastBoot, but only if they've been
+explicitly whitelisted first.
+
+To whitelist a dependency, you'll need to edit your addon's
+`package.json`. You probably already have an `ember-addon` field; to
+this hash, add a property called `fastBootDependencies` that contains an
+array of Node modules that may be used.
+
+Make sure that any modules you want to access are also included in your
+`package.json`'s `dependencies` field.
+
+For example, if I have an addon where I need to use both the built-in
+`path` module as well as `redis`, my `package.json` might look like
+this:
+
+```js
+{
+  "name": "my-ember-addon",
+  "version": "1.0.0",
+  // ...
+  "dependencies": {
+    "redis": "^2.6.0"
+  },
+  "ember-addon": {
+    "configPath": "tests/dummy/config",
+    "fastBootDependencies": [
+      "path",
+      "redis"
+    ]
+  }
+}
+```
+
+Because `path` is built-in, I don't need to add it to my dependencies.
+But because `redis` comes from npm, I need to add it as a dependency. In
+both cases, I must explicitly whitelist that both are available to the
+Ember app.
+
+Now that we've configured `path` and `redis` to be available from
+within our Ember app, addon code can require it using the
+`FastBoot.require()` method:
+
+```js
+const redis = FastBoot.require('redis');
+
+let client = redis.createClient();
+```
+
+Note that the `FastBoot` global is _only_ available when running in
+FastBoot and won't exist when your app is running in the browser. If you
+have some code that runs in both environments, make sure to check
+whether `FastBoot` exists before you access it. (And, of course, someone
+may be using your addon in an app that doesn't have FastBoot installed.)
+
+## Accessing the FastBoot Service
+
+As discussed in the [User Guide](user-guide), you can access information
+about the current request via the `fastboot` service, accessible like
+any other service in an Ember app.
+
+One thing to note is that someone may use your addon in an app that
+doesn't have FastBoot installed. In that case, if your addon tries to
+inject the `fastboot` service, they'll get an exception saying the
+`fastboot` service cannot be found.
+
+Instead, you can write a computed property that uses the low-level
+`getOwner` functionality to lookup the `fastboot` service directly:
+
+```js
+import getOwner from "ember-owner/get";
+import Service from "ember-service";
```

Comment:
@locks ember-cli-shims has been shipping by default in Ember CLI apps for… over a year now? At this point we definitely can't break them.